### ISSUE TITLE
Make it work nicely with wayland

### ISSUE THREAD

**Issue by Atemu** [kreijack/xlibinput_calibrator#9]
Running a calibration on wayland works and it gets applied but only half way. It prints `WARNING: xinput is running against an Xwayland server. See the xinput man page for details.` which is true; it's running against xwayland and that means calibration only applies to xwayland apps. Switch to a native wayland app and the stylus is no longer calibrated.

The issue is twofold:

1. We need to use a sane, portable method to apply the transformation matrix at runtime for Wayland compositors. You probably need a protocol for this; perhaps something like that already exists.
2. The app should suggest the user a means of permanently applying the matrix. Under X11, that's the xorg config files but under Wayland its...? Udev?

**Comment by kreijack** [comment]
I am curious how you get the matrix in wayland; but this is another history.

Looking at the code of libinput, it seems that there are two way to set the calibration matrix ( see also [libinput udev config](https://wayland.freedesktop.org/libinput/doc/1.8.2/udev_config.html) )
1) using the function [libinput_device_config_calibration_set_matrix()](https://wayland.freedesktop.org/libinput/doc/1.8.2/group__config.html#ga09a798f58cc601edd2797780096e9804), but this is valid only inside the application (i.e. the wayland compositor which compute the pointer coordinates). It is not doable without updating the code of the compositor.
2) "attaching" the coordinate to the device using udev, something like (from [Stackoverflow - touch calibration](https://stackoverflow.com/questions/67162466/where-is-the-calibration-matrix-in-weston-wayland) )
```
echo 'SUBSYSTEM=="input", KERNEL=="event[0-9]*", ENV{ID_INPUT_TOUCHSCREEN}=="1", ENV{LIBINPUT_CALIBRATION_MATRIX}="'c1 c2 c3 c4 c5 c6'"' >> /etc/udev/rules.d/touchscreen.rules
```

The compositor libinput, when launched, loads these information and it sets according the calibration matrix; the `c1 c2 c3 c4 c5 c6` are the matrix coeffiecient.

Also [here](https://www.mail-archive.com/wayland-devel@lists.freedesktop.org/msg41436.html) you can see more information.

What we can do:
1) updating xlibinput_calibrator to support wayland is not easy, because wayland prevent to open a window in a "full screen mode". So for now we need to continue run xlibinput_calibrator in Xorg. If you was successful to run it in Xwayland, please could you share how you did ?
2) I can add a new option "--show-udev-setting" to show a a command like the one above to set the LIBINPUT_CALIBRATION_MATRIX

**Comment by Atemu** [comment]
> I am curious how you get the matrix in wayland

I didn't. I only got the matrix to apply in Xwayland via xinput.

> If you was successful to run it in Xwayland, please could you share how you did ?

Just ran the app and it works for all Xwayland windows.

>     I can add a new option "--show-udev-setting" to show a a command like the one above to set the LIBINPUT_CALIBRATION_MATRIX

A Udev rule is the only way I've found that *should* work. That'd solve point 2.

On a side note, the LIBINPUT_CALIBRATION_MATRIX does work for my touch screen but not my stylus :/

https://gitlab.freedesktop.org/libinput/libinput/-/issues/891

**Comment by kreijack** [comment]
> >    If you was successful to run it in Xwayland, please could you share how you did ?
> Just ran the app and it works for all Xwayland windows.

Yes because Xorg has a global calibration that it is accessible via xinput.


I don't own a Wacom tablet, so what will suggest it is only a try. However the above udev rule is for touchscreen (look at ID_INPUT_TOUCHSCREEN==1 ). Mind to replace this with ID_INPUT_TABLET==1 ? E.g.:
```
echo 'SUBSYSTEM=="input", KERNEL=="event[0-9]*", ENV{ID_INPUT_TABLET}=="1", ENV{LIBINPUT_CALIBRATION_MATRIX}="'c1 c2 c3 c4 c5 c6'"' >> /etc/udev/rules.d/touchscreen.rules
```
Where the  `c1 c2 c3 c4 c5 c6` are the matrix coeffiecient.

Other possible filter are (from [udev_config](https://wayland.freedesktop.org/libinput/doc/1.7.0/udev_config.html) ) :

ID_INPUT_KEYBOARD
ID_INPUT_KEY
 ID_INPUT_MOUSE
 ID_INPUT_TOUCHPAD
 ID_INPUT_TOUCHSCREEN
 ID_INPUT_TABLET
 ID_INPUT_JOYSTICK
 ID_INPUT_ACCELEROMETER

**Comment by Atemu** [comment]
As mentioned in the libinput issue, I did not apply any filters to my rule and the matrix shows up when querying libinput devices.

**Comment by kreijack** [comment]
Ok, I didn't read carefully your libinput issue.

However my understanding is that a tablet may not be scaled. My hypothesis is that a touchscreen has a 1:1 relation to a physical screen. So a global scale makes sense. However a tablet is a different beast that usually  has a relation to a window of an application: tablet with a pen is used to "drive" a brush inside a window which may be stretched in several ways.

I saw in your libinput issue that there are two Wacom devices: are these two distinct physical device or it is the same physical device that appears as two logical device ?

Anyway I saw that in the libinput source the application of a calibration matrix is conditioned by the INPUT_PROP_DIRECT property. And this property is set depending by the wacom device (see linux/driver/hid/wacom_wac.c
```
void wacom_setup_device_quirks(struct wacom *wacom)
{
[...]
        switch (features->type) {
        case PL:
        case DTU:
        case DTUS:
        case DTUSX:
        case WACOM_21UX2:
        case WACOM_22HD:
        case DTK:
        case WACOM_24HD:
        case WACOM_27QHD:
        case CINTIQ_HYBRID:
        case CINTIQ_COMPANION_2:
        case CINTIQ:
        case WACOM_BEE:
        case WACOM_13HD:
        case WACOM_24HDT:
        case WACOM_27QHDT:
        case TABLETPC:
        case TABLETPCE:
        case TABLETPC2FG:
        case MTSCREEN:
        case MTTPC:
        case MTTPC_B:
                features->device_type |= WACOM_DEVICETYPE_DIRECT;
                break;
        }
[...]
int wacom_setup_touch_input_capabilities(struct input_dev *input_dev,
                                         struct wacom_wac *wacom_wac)
{
[...]
        if (features->device_type & WACOM_DEVICETYPE_DIRECT)
                __set_bit(INPUT_PROP_DIRECT, input_dev->propbit);

```

So maybe this is the explanation why you a calibration matrix is ignore for certain device.

Could you share the exact Wacom model ?

**Comment by Atemu** [comment]
Could we continue that conversation in the libinput issue? It's a bit off-topic here I feel. Everything past the matrix application is a libinput or kernel issue, not one of libinput_calibrator.

The application of the matrix via udev seems to work alright and it does indeed work for my touch screen, so I think libinput_calibrator should be recommending that.

It'd be great if you could make the recommendation printing configurable at build time such that distros can recommend their own internal methods of configuring udev rules and/or put a link to the docs.  
On NixOS, we'd want to recommend the user to put the udev rule into `services.udev.extraRules` and/or link them to https://search.nixos.org/options?query=services.udev.extraRules&show=services.udev.extraRules for instance.
